core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java

```diff
@@ -125,137 +120,131 @@ public GroupByNumericArrayIterator(final QueryNode node,
     }
-    
+
     executor = node.pipelineContext().tsdb().quickWorkPool();
-    
+
     this.result = (GroupByResult) result;
-    final NumericArrayAggregatorFactory factory = node.pipelineContext().tsdb()
-        .getRegistry().getPlugin(NumericArrayAggregatorFactory.class, 
-            ((GroupByConfig) node.config()).getAggregator());
+    final NumericArrayAggregatorFactory factory =
+        node.pipelineContext()
+            .tsdb()
+            .getRegistry()
+            .getPlugin(
+                NumericArrayAggregatorFactory.class,
+                ((GroupByConfig) node.config()).getAggregator());
     if (factory == null) {
-      throw new IllegalArgumentException("No aggregator factory found of type: " 
-          + ((GroupByConfig) node.config()).getAggregator());
-    }
-    aggregator = factory.newAggregator(((GroupByConfig) node.config()).getInfectiousNan());
-    if (aggregator == null) {
-      throw new IllegalArgumentException("No aggregator found of type: " 
-          + ((GroupByConfig) node.config()).getAggregator());
+      throw new IllegalArgumentException(
+          "No aggregator factory found of type: "
+              + ((GroupByConfig) node.config()).getAggregator());
     }
-    
-    long startTs = DateTime.nanoTime();
-    
-    TemporalAmount temporalAmount = this.result.downstreamResult().timeSpecification().interval();
-    
-    TimeStamp ts = node.pipelineContext().query().startTime().getCopy();
-    int intervals = 0;
-    while (ts.compare(Op.LT, node.pipelineContext().query().endTime())) {
-      intervals++;
-      ts.add(temporalAmount);
+
+    int size;
+    DownsampleConfig downsampleConfig = ((GroupBy) node).getDownsampleConfig();
+    if (null == downsampleConfig) {
+      TemporalAmount temporalAmount = this.result.downstreamResult().timeSpecification().interval();
+      TimeStamp ts = node.pipelineContext().query().startTime().getCopy();
+      int intervals = 0;
+      while (ts.compare(Op.LT, node.pipelineContext().query().endTime())) {
+        intervals++;
+        ts.add(temporalAmount);
+      }
+      size = intervals;
+    } else {
+      size = downsampleConfig.intervals();
     }
-    
-    logger.info("Sources size {}, Intervals found {}", sources.size(), intervals);
 
-    valuesCombiner = new NumericArrayAggregator[NUM_THREADS];
-    for (int i = 0; i < valuesCombiner.length; i++) {
-      valuesCombiner[i] = factory.newAggregator(((GroupByConfig) node.config()).getInfectiousNan());
+    logger.trace("Group size: " + size);
+
+    aggregator =
+        factory.newAggregator(((GroupByConfig) node.config()).getInfectiousNan());
+    if (aggregator == null) {
+      throw new IllegalArgumentException(
+          "No aggregator found of type: " + ((GroupByConfig) node.config()).getAggregator());
     }
 
     // TODO: Need to check if it makes sense to make this threshold configurable
     if (sources.size() >= NUM_THREADS) {
```

Comment:
Could probably do some factor of NUM_THREADS like `NUM_THREADS * 16`

Reply:
Remember, I commented on GSK's pr to make it configurable. This one is on you @manolama